From: Liam Howlett <Liam.Howlett@oracle.com>
Subject: mm/madvise: fix VMA_ITERATOR start position
Date: Wed, 25 Jan 2023 08:58:09 -0500

The WARN_ON() in vma_iter_store() detected an invalid VMA iterator state. 
Inspecting the code stack from the report shows that the VMA iterator is
never set to the correct start position.

Setting the initialization of the VMA iterator to use the address 'start'
fixes this issue.

Link: https://lkml.kernel.org/r/20230125135809.85262-1-Liam.Howlett@oracle.com
Signed-off-by: Liam Howlett <Liam.Howlett@oracle.com>
Reported-by: <syzbot+1d9ec4d1a334ae9e1ca6@syzkaller.appspotmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/madvise.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/mm/madvise.c~madvise-use-split_vma-instead-of-__split_vma-fix
+++ a/mm/madvise.c
@@ -142,7 +142,7 @@ static int madvise_update_vma(struct vm_
 	struct mm_struct *mm = vma->vm_mm;
 	int error;
 	pgoff_t pgoff;
-	VMA_ITERATOR(vmi, mm, 0);
+	VMA_ITERATOR(vmi, mm, start);
 
 	if (new_flags == vma->vm_flags && anon_vma_name_eq(anon_vma_name(vma), anon_name)) {
 		*prev = vma;
_
